### PR TITLE
5.10.x: PTP_KVM support for arm/arm64

### PR DESCRIPTION
**Pull request by damon-kwok**
The enablement of ptp_kvm for arm is completed, see [[1]](https://lore.kernel.org/kvmarm/20210330145430.996981-1-maz@kernel.org/).
It's for `kernel 5.13.x`, so we have to apply private patch to enable it in `kernel 5.10`.
The `ptp_kvm` can offer the capability of time sync in `kata` even there
is no network available and higher precision than time sync service depend on network.

In order for the `ptp_kvm` module to work on all kernel versions, I did some work. 
I merged the code from  `kernel 5.13`, and made patches for different versions.
[patch-1](https://github.com/kata-containers/kata-containers/pull/2105) for `kernel 4.19.x LTS`
[patch-2](https://github.com/kata-containers/kata-containers/pull/2058) for `kernel 5.4.x LTS`
[patch-3](https://github.com/kata-containers/kata-containers/pull/2105) for `kernel 5.10.x LTS`

Test:
|HOST  |  GUEST |  RESULT|
| --- | -------|---- |
|4.19   |  4.19   |  pass|
|5.4   |   5.4    |  pass|
|5.10   |  5.10   |  pass|
|4.19   |  5.4    |  pass|
|4.19  |   5.10   |  pass|
|5.4   |   4.19   |  pass|
|5.4    |  5.10   |  pass|
|5.10  |   5.4    |  pass|
|5.10  |   4.19   |  pass|
|5.12  |   4.19   |  pass|
|4.19  |   5.12   |  pass|

```sh
[root@localhost ~]# chronyc sources
210 Number of sources = 1
MS Name/IP address         Stratum Poll Reach LastRx Last sample               
===============================================================================
#* PHC0                          0   2   377     4     +1ns[   +1ns] +/-   60ns
```
---
- [1] https://lore.kernel.org/kvmarm/20210330145430.996981-1-maz@kernel.org/

### CODE REVIEW

**Comment by fidencio** [comment]
@jongwu, may I bother you for a review on this one?

File: tools/packaging/kernel/patches/5.10.x/0001-arm64-Add-host-guest-KVM-PTP-support-in-kata.patch [hunk 1]
```diff
@@ -0,0 +1,985 @@
...
+-	if (!hv_clock)
+-		return -ENODEV;
+-
+-	ret = kvm_hypercall2(KVM_HC_CLOCK_PAIRING, clock_pair_gpa,
+-			KVM_CLOCK_PAIRING_WALLCLOCK);
+-	if (ret == -KVM_ENOSYS || ret == -KVM_EOPNOTSUPP)
+-		return -ENODEV;
++	ret = kvm_arch_ptp_init();
++	if (ret) {
++		pr_err("fail to initialize ptp_kvm");
```
Comment (jongwu) [inline tools/packaging/kernel/patches/5.10.x/0001-arm64-Add-host-guest-KVM-PTP-support-in-kata.patch:652]: please change this as following:
if (ret) {
-		pr_err("fail to initialize ptp_kvm");
+		if (ret != -EOPNOTSUPP)
+			pr_err("fail to initialize ptp_kvm");

File: tools/packaging/kernel/patches/5.10.x/0001-arm64-Add-host-guest-KVM-PTP-support-in-kata.patch [hunk 2]
```diff
@@ -0,0 +1,985 @@
...
++PTP_KVM support for arm/arm64
++=============================
++
++PTP_KVM is used for time sync between guest and host in a high precision.
++It needs to get the wall time and counter value from the host and transfer these
++to guest via hypercall service. So one more hypercall service has been added.
++
++This new SMCCC hypercall is defined as:
++
++* ARM_SMCCC_HYP_KVM_PTP_FUNC_ID: 0x86000001
```
Comment (jongwu) [inline tools/packaging/kernel/patches/5.10.x/0001-arm64-Add-host-guest-KVM-PTP-support-in-kata.patch:81]: change this to ARM_SMCCC_VENDOR_HYP_KVM_PTP_FUNC_ID

File: tools/packaging/kernel/patches/5.10.x/0001-arm64-Add-host-guest-KVM-PTP-support-in-kata.patch [hunk 3]
```diff
@@ -0,0 +1,985 @@
...
++
++As both 32 and 64-bits ptp_kvm client should be supported, we choose SMC32/HVC32
++calling convention.
++
++ARM_SMCCC_HYP_KVM_PTP_FUNC_ID:
```
Comment (jongwu) [inline tools/packaging/kernel/patches/5.10.x/0001-arm64-Add-host-guest-KVM-PTP-support-in-kata.patch:86]: change this to ARM_SMCCC_VENDOR_HYP_KVM_PTP_FUNC_ID

File: tools/packaging/kernel/patches/5.10.x/0001-arm64-Add-host-guest-KVM-PTP-support-in-kata.patch [hunk 4]
```diff
@@ -0,0 +1,985 @@
...
++ * guest can touch it.
++ */
++#include <linux/kernel.h>
++#include <linux/err.h>
++static inline bool kvm_arm_hyp_service_available(u32 func_id)
++{
++	extern DECLARE_BITMAP(__kvm_arm_hyp_services, ARM_SMCCC_KVM_NUM_FUNCS);
++
++	if (func_id >= ARM_SMCCC_KVM_NUM_FUNCS)
++		return -EINVAL;
```
Comment (jongwu) [inline tools/packaging/kernel/patches/5.10.x/0001-arm64-Add-host-guest-KVM-PTP-support-in-kata.patch:838]: please return false here.

**Comment by jongwu** [comment]
Thanks @damon-kwok -, I give some inline comments. Also I'm interested with another thing. As you have applied ptp_kvm patch on kernel 5.10, 5.4 and 4.19, I wonder if they can work well when the guest kernel and host kernel are in different version, e.g. run 5.10 guest in 4.19 host. I think it very meaningful to let it work.        `

Another small comment, please add the reference of this patch.

**Comment by damon-kwok** [comment]
@jongwu said:
> Also I'm interested with another thing. As you have applied ptp_kvm patch on kernel 5.10, 5.4 and 4.19, I wonder if they can work well when the guest kernel and host kernel are in different version, e.g. run 5.10 guest in 4.19 host. I think it very meaningful to let it work. `

This is indeed a problem. I read the v19 patch source code. We need to make changes to the three patches 4.19.x, 5.4.x and 5.10.x at the same time. I will work for this.

**Comment by damon-kwok** [comment]
@jodh-intel @jongwu @Pennyzct  @justin-he
In order for the `ptp_kvm` module to work on all kernel versions, I did some work. 
I merged the code from  `kernel 5.13`, and made patches for different versions.
#2058 #2063 #2105

**Comment by jongwu** [comment]
Hello @damon-kwok -, I'm very appreciate of your great work! You have done what I would have to do.  It's really a lot of work to do so much test between different version of kernel. You just do more than I expected! Thanks @damon-kwok !
Since you have tested these patches, I will just have a quick look of them later.

File: tools/packaging/kernel/patches/5.10.x/0001-5.10.x-PTP_KVM-support-for-arm-arm64-in-Kata.patch [hunk 5]
```diff
@@ -438,20 +440,20 @@ index 00c88b809..e153c71ec 100644
...
+index deb429a..5523f96 100644
 --- a/drivers/ptp/Kconfig
 +++ b/drivers/ptp/Kconfig
 @@ -108,7 +108,7 @@ config PTP_1588_CLOCK_PCH
  config PTP_1588_CLOCK_KVM
  	tristate "KVM virtual PTP clock"
  	depends on PTP_1588_CLOCK
 -	depends on KVM_GUEST && X86
-+	depends on KVM_GUEST && X86 || (HAVE_ARM_SMCCC_DISCOVERY && ARM_ARCH_TIMER)
++	depends on X86 || (HAVE_ARM_SMCCC_DISCOVERY && ARM_ARCH_TIMER)
```
Comment (jongwu) [inline tools/packaging/kernel/patches/5.10.x/0001-5.10.x-PTP_KVM-support-for-arm-arm64-in-Kata.patch:451]: why remove "KVM_GUEST" here? it may impact x86.

**Comment by jongwu** [comment]
Hi @damon-kwok -, please add "signed-off-by" and "Fixes" segments in your commit, [here](https://github.com/kata-containers/kata-containers/pull/2027/commits/87536dc33b13a2fcdc12a20e967abfc212c0d1ea) is a reference.

**Pull request closed without merging**